File: sdk-java/hfcx-sdk-client/src/main/java/eg/gov/healthflow/hfcx/sdk/client/recipient/LocalKeyProvider.java
package eg.gov.healthflow.hfcx.sdk.client.recipient;

import java.security.interfaces.RSAPrivateKey;

/**
 * Source of the recipient's RSA private key. Pluggable so participants
 * can ship their own implementations against AWS KMS, Azure Key Vault,
 * an HSM, etc. The SDK ships two reference implementations:
 *
 * <ul>
 *   <li>{@link FileLocalKeyProvider} — reads PEM from a filesystem path.</li>
 *   <li>{@link VaultLocalKeyProvider} — reads from HashiCorp Vault KV v2.</li>
 * </ul>
 *
 * <p>Implementations MUST NOT cache the key bytes on disk. The
 * {@link RSAPrivateKey} object is the only material that should ever
 * leave this interface.
 */
@FunctionalInterface
public interface LocalKeyProvider {

    /**
     * @return the recipient's current RSA private key.
     */
    RSAPrivateKey getPrivateKey();
}

File: sdk-javascript/src/recipient/LocalKeyProvider.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import type { KeyLike } from 'jose';

/**
 * Abstraction over the recipient's RSA private key source. Sister to
 * Java's `LocalKeyProvider`, Python's `LocalKeyProvider` Protocol, and
 * .NET's `ILocalKeyProvider`.
 *
 * Two reference implementations ship with the SDK:
 * {@link FileLocalKeyProvider} for PKCS#8 PEM on disk and
 * {@link VaultLocalKeyProvider} for HashiCorp Vault KV v2.
 */
export interface LocalKeyProvider {
  /**
   * Return the recipient's RSA private key. May be called once per
   * inbound request — implementations should be cheap and side-effect-
   * free, but file/Vault re-reads on every call ARE the contract so
   * key rotations take effect immediately.
   *
   * @throws {KeyUnavailableError} if the key cannot be loaded.
   */
  getPrivateKey(): Promise<KeyLike>;
}
